x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/ApiKeyService.java

```diff
@@ -332,6 +353,60 @@ private void createApiKeyAndIndexIt(
         }));
     }
 
+    public void updateApiKey(
+        final Authentication authentication,
+        final UpdateApiKeyRequest request,
+        final Set<RoleDescriptor> userRoles,
+        final ActionListener<UpdateApiKeyResponse> listener
+    ) {
+        ensureEnabled();
+
+        if (authentication == null) {
+            listener.onFailure(new IllegalArgumentException("authentication must be provided"));
+            return;
+        } else if (authentication.isApiKey()) {
+            listener.onFailure(new IllegalArgumentException("authentication via an API key is not supported for updating API keys"));
+            return;
+        }
+
+        logger.debug("Updating API key [{}]", request.getId());
+
+        findVersionedApiKeyDocsForSubject(authentication, new String[] { request.getId() }, ActionListener.wrap((versionedDocs) -> {
+            final var apiKeyId = request.getId();
+
+            if (versionedDocs.isEmpty()) {
+                throw new ResourceNotFoundException("no API key owned by requesting user found for ID [" + apiKeyId + "]");
+            }
+
+            validateCurrentApiKeyDocForUpdate(apiKeyId, authentication, single(apiKeyId, versionedDocs).doc());
+
+            executeBulkRequest(
+                buildBulkRequestForUpdate(versionedDocs, authentication, request, userRoles),
+                ActionListener.wrap(bulkResponse -> translateResponseAndClearCache(apiKeyId, bulkResponse, listener), listener::onFailure)
+            );
+        }, listener::onFailure));
+    }
+
+    // package-private for testing
+    void validateCurrentApiKeyDocForUpdate(String apiKeyId, Authentication authentication, ApiKeyDoc apiKeyDoc) {
+        assert authentication.getEffectiveSubject().getUser().principal().equals(apiKeyDoc.creator.get("principal"));
+
+        if (apiKeyDoc.invalidated) {
+            throw new IllegalArgumentException("cannot update invalidated API key [" + apiKeyId + "]");
+        }
+
+        final var now = clock.instant();
+        boolean expired = apiKeyDoc.expirationTime != -1
+            && (Instant.ofEpochMilli(apiKeyDoc.expirationTime).equals(now) || Instant.ofEpochMilli(apiKeyDoc.expirationTime).isBefore(now));
```

Comment:
Can this be 
```suggestion
            && now.isAfter(Instant.ofEpochMilli(apiKeyDoc.expirationTime));
```
?

Reply:
Yup. I went for the other option to keep the "before" object to the left, but this is cleaner because it gets rid of the `||`. Thanks!